File: R/fa2.R
# forceatlas2 ------------------------------------------------------------------

## param combination -----------------------------------------------------------

#' Internal helper to prepare the ForceAtlas2 parameters
#'
#' @param knn_method String. Method to use to generate the kNN graph.
#' @param nn_params Named list. The nearest neighbour search parameters.
#' @param fa2_params Named list. The ForceAtlas2-specific parameters.
#'
#' @return Returns the list of final parameters.
#'
#' @export
#'
#' @keywords internal
.prepare_fa2_params <- function(knn_method, nn_params, fa2_params) {
  # checks
  checkmate::assertChoice(
    knn_method,
    c(
      "kmknn",
      "balltree",
      "hnsw",
      "annoy",
      "nndescent",
      "exhaustive",
      "ivf"
    )
  )
  assertNnParams(nn_params)
  assertFa2Params(fa2_params)

  final_params <- c(nn_params, fa2_params)
  final_params[["knn_method"]] <- knn_method

  final_params
}

## main functions --------------------------------------------------------------

#' Rust-based ForceAtlas2
#'
#' @description Lays out the kNN graph of the data with ForceAtlas2, the way
#' scanpy's `draw_graph` does: the kNN graph is turned into the UMAP fuzzy
#' union graph and every edge pulls, every node pair pushes (`1/d`, with masses
#' `1 + degree`), plus gravity towards the origin. Repulsion runs on a
#' Barnes-Hut tree. Already have a graph? Use [forceatlas2_from_graph()].
#'
#' @param data Numerical matrix or data frame. The data to embed of shape
#' samples x features. Will be coerced to a matrix.
#' @param knn Optional `NearestNeighbours` class. If provided, the k-nearest
#' neighbour search is skipped and this one is used. Defaults to `NULL`.
#' @param k Integer. Number of nearest neighbours. Defaults to `15L`.
#' @param knn_method Character. (Approximate) Nearest neighbour method to use.
#' One of `"kmknn"`, `"hnsw"`, `"annoy"`, `"nndescent"`, `"balltree"`, `"ivf"`
#' or `"exhaustive"`. Defaults to `"kmknn"`.
#' @param nn_params Named list. Nearest neighbour search parameters, see
#' [params_nn()].
#' @param fa2_params Named list. ForceAtlas2 parameters, see [params_fa2()].
#' @param seed Integer. Random seed for reproducibility. Defaults to `42L`.
#' @param use_high_precision Optional boolean. Gives fine-grained control over
#' `fp32` vs `fp64` usage.
#' @param .verbose Logical. Controls verbosity. Defaults to `TRUE`.
#'
#' @return A numerical matrix with dimensions samples x 2 containing the
#' ForceAtlas2 layout.
#'
#' @references Jacomy, et al., PLoS ONE, 2014
#'
#' @export
forceatlas2 <- function(
  data,
  knn = NULL,
  k = 15L,
  knn_method = c(
    "kmknn",
    "balltree",
    "hnsw",
    "annoy",
    "nndescent",
    "exhaustive",
    "ivf"
  ),
  nn_params = params_nn(),
  fa2_params = params_fa2(),
  seed = 42L,
  use_high_precision = NULL,
  .verbose = TRUE
) {
  # transformation
  if (is.data.frame(data)) {
    data <- as.matrix(data)
  }
  knn_method <- match.arg(knn_method)

  checkmate::assert_matrix(
    data,
    mode = "numeric",
    any.missing = FALSE,
    min.rows = 2,
    min.cols = 1
  )
  checkmate::assert(
    checkmate::testNull(knn),
    checkmate::testClass(knn, "NearestNeighbours")
  )
  checkmate::qassert(k, "I1[2,)")
  checkmate::qassert(.verbose, c("B1", "I1[0, 2]"))
  checkmate::qassert(use_high_precision, c("0", "B1"))
  checkmate::qassert(seed, "I1")

  final_fa2_params <- .prepare_fa2_params(
    knn_method = knn_method,
    nn_params = nn_params,
    fa2_params = fa2_params
  )

  res <- if (!is.null(knn)) {
    if (.verbose) {
      message("Using provided kNN graph.")
    }
    rs_forceatlas2_from_knn(
      embd = data,
      knn_data = knn,
      k = k,
      fa2_params = final_fa2_params,
      seed = seed,
      use_high_precision = use_high_precision,
      verbose = parse_verbosity(.verbose)
    )
  } else {
    rs_forceatlas2(
      embd = data,
      k = k,
      fa2_params = final_fa2_params,
      seed = seed,
      use_high_precision = use_high_precision,
      verbose = parse_verbosity(.verbose)
    )
  }

  res
}

#' Rust-based ForceAtlas2 on a pre-computed graph
#'
#' @description Lays out any undirected graph with ForceAtlas2, for example an
#' SNN graph or a graph from another package. Edge weights come from the
#' `weight` edge attribute; without one, every edge weighs `1`. Self-loops are
#' dropped. Only the optimisation parameters in `fa2_params` are used; the
#' graph and initialisation ones do not apply here.
#'
#' @param graph An undirected `igraph` object without multi-edges. Use
#' `igraph::as_undirected(mode = "collapse")` or `igraph::simplify()` to get
#' there. If present, the `weight` edge attribute needs to be finite and
#' non-negative.
#' @param init Optional numerical matrix of dimensions `vcount(graph) x 2`.
#' The initial layout, e.g. a UMAP of the same samples. If `NULL`, a random
#' layout is used. Defaults to `NULL`.
#' @param fa2_params Named list. ForceAtlas2 parameters, see [params_fa2()].
#' @param seed Integer. Random seed for reproducibility. Defaults to `42L`.
#' @param use_high_precision Optional boolean. Gives fine-grained control over
#' `fp32` vs `fp64` usage.
#' @param .verbose Logical. Controls verbosity. Defaults to `TRUE`.
#'
#' @return A numerical matrix with dimensions `vcount(graph) x 2` containing
#' the ForceAtlas2 layout, rows in vertex order.
#'
#' @references Jacomy, et al., PLoS ONE, 2014
#'
#' @export
forceatlas2_from_graph <- function(
  graph,
  init = NULL,
  fa2_params = params_fa2(),
  seed = 42L,
  use_high_precision = NULL,
  .verbose = TRUE
) {
  checkmate::assertClass(graph, "igraph")
  if (igraph::is_directed(graph)) {
    stop(
      "ForceAtlas2 needs an undirected graph. ",
      "Use igraph::as_undirected(graph, mode = \"collapse\")."
    )
  }
  if (igraph::any_multiple(graph)) {
    stop(
      "The graph has multi-edges. Use igraph::simplify() to merge them."
    )
  }
  n <- igraph::vcount(graph)
  checkmate::assertNumber(n, lower = 2)
  checkmate::assert(
    checkmate::testNull(init),
    checkmate::testMatrix(
      init,
      mode = "numeric",
      any.missing = FALSE,
      nrows = n,
      ncols = 2
    )
  )
  assertFa2Params(fa2_params)
  checkmate::qassert(.verbose, c("B1", "I1[0, 2]"))
  checkmate::qassert(use_high_precision, c("0", "B1"))
  checkmate::qassert(seed, "I1")

  if (!is.null(init)) {
    storage.mode(init) <- "double"
  }

  edges <- igraph::as_edgelist(graph, names = FALSE)
  weight <- igraph::E(graph)$weight
  if (is.null(weight)) {
    weight <- rep(1, nrow(edges))
  }
  checkmate::assertNumeric(
    weight,
    lower = 0,
    finite = TRUE,
    any.missing = FALSE
  )

  rs_forceatlas2_from_graph(
    from = as.integer(edges[, 1]),
    to = as.integer(edges[, 2]),
    weight = as.numeric(weight),
    n = as.integer(n),
    init = init,
    fa2_params = fa2_params,
    seed = seed,
    use_high_precision = use_high_precision,
    verbose = parse_verbosity(.verbose)
  )
}

File: src/rust/src/fa2.rs
//! ForceAtlas2 wrapper functions to R from manifolds-rs

#![warn(missing_docs)]

use ann_search_rs::cpu::hnsw::{HnswIndex, HnswState};
use ann_search_rs::cpu::nndescent::{NNDescent, NNDescentQuery};
use ann_search_rs::utils::nndescent_utils::ApplySortedUpdates;
use bixverse_rs::prelude::IntoExtendrErr;
use extendr_api::{List, Robj};
use faer::{Mat, MatRef};
use manifolds_rs::prelude::*;
use manifolds_rs::*;
use rand_distr::{Distribution, StandardNormal};
use std::collections::HashMap;

use crate::umap::get_params_umap_graph;
use crate::utils::get_params_nn_manifolds;

/// Repulsion approximation passed to manifolds-rs. Barnes-Hut is the only
/// variant the crate offers, so it is not exposed to R.
const FA2_APPROX: &str = "barnes_hut";

////////////
// Params //
////////////

/// Helper function to generate the ForceAtlas2 optimisation parameters
///
/// ### Params
///
/// * `r_list` - The list that has the ForceAtlas2 optimisation parameters.
///
/// ### Returns
///
/// The `Fa2OptimParams` with the crate defaults for anything not found in the
/// list.
fn get_params_fa2_optim<T>(r_list: List) -> Result<Fa2OptimParams<T>, extendr_api::Error>
where
    T: ManifoldsFloat,
{
    let optim_params: HashMap<&str, Robj> = r_list.try_into()?;
    let defaults = Fa2OptimParams::<T>::default();

    let get_real = |name: &str, default: T| -> T {
        optim_params
            .get(name)
            .and_then(|v| v.as_real())
            .map(|v| T::from_f64(v).unwrap())
            .unwrap_or(default)
    };
    let get_bool = |name: &str, default: bool| -> bool {
        optim_params
            .get(name)
            .and_then(|v| v.as_bool())
            .unwrap_or(default)
    };

    let n_epochs = optim_params
        .get("n_epochs")
        .and_then(|v| v.as_integer())
        .map(|v| v as usize)
        .unwrap_or(defaults.n_epochs);

    Ok(Fa2OptimParams::new(
        n_epochs,
        get_real("scaling_ratio", defaults.scaling_ratio),
        get_real("gravity", defaults.gravity),
        get_bool("strong_gravity", defaults.strong_gravity),
        get_bool("lin_log", defaults.lin_log),
        get_bool("dissuade_hubs", defaults.dissuade_hubs),
        get_real("edge_weight_influence", defaults.edge_weight_influence),
        get_real("jitter_tolerance", defaults.jitter_tolerance),
        get_real("theta", defaults.theta),
    ))
}

/// Helper function to assemble the `Fa2Params` from an R list
///
/// `mix_weight` is never sent from R, so `get_params_umap_graph` falls back to
/// `1`, the only value `forceatlas2()` accepts.
///
/// ### Params
///
/// * `fa2_params` - Named R list with the nearest neighbour, graph and
///   ForceAtlas2 parameters.
/// * `k` - Number of neighbours to use.
///
/// ### Returns
///
/// The assembled `Fa2Params`.
fn build_fa2_params<T>(fa2_params: List, k: usize) -> Result<Fa2Params<T>, extendr_api::Error>
where
    T: ManifoldsFloat,
{
    let nn_params = get_params_nn_manifolds(fa2_params.clone())?;
    let graph_params = get_params_umap_graph(fa2_params.clone())?;
    let optim_params = get_params_fa2_optim(fa2_params.clone())?;

    let params: HashMap<&str, Robj> = fa2_params.try_into()?;

    let get_string = |name: &str, default: &str| -> String {
        String::from(
            params
                .get(name)
                .and_then(|v| v.as_str())
                .unwrap_or(default),
        )
    };

    let randomised = params
        .get("randomised")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);

    Ok(Fa2Params::new(
        k,
        get_string("knn_method", "kmknn"),
        get_string("init", "spectral"),
        None,
        randomised,
        nn_params,
        graph_params,
        optim_params,
    ))
}

/////////////////
// ForceAtlas2 //
/////////////////

/// Wrapper function into the ForceAtlas2 implementation in `manifolds-rs`
///
/// Builds the UMAP fuzzy union graph from the data (or the pre-computed kNN)
/// and lays it out with ForceAtlas2.
///
/// ### Params
///
/// * `data` - The data to embed, samples x features.
/// * `pre_computed_knn` - Optional pre-computed kNN to be used.
/// * `k` - Number of neighbours to use.
/// * `fa2_params` - Named R list with the nearest neighbour, graph and
///   ForceAtlas2 parameters.
/// * `seed` - For reproducibility.
/// * `verbose` - If `0` -> silent or `1` for normal verbosity, `2` for detailed
///   verbosity.
///
/// ### Returns
///
/// The ForceAtlas2 embedding as samples x 2 matrix.
///
/// ### References
///
/// Jacomy et al., PLoS ONE, 2014
pub fn fa2_manifold<T>(
    data: MatRef<T>,
    pre_computed_knn: PreComputedKnn<T>,
    k: usize,
    fa2_params: List,
    seed: usize,
    verbose: usize,
) -> Result<Mat<T>, extendr_api::Error>
where
    T: ManifoldsFloat,
    HnswIndex<T>: HnswState<T>,
    NNDescent<T>: ApplySortedUpdates<T> + NNDescentQuery<T>,
    StandardNormal: Distribution<T>,
{
    let fa2_params = build_fa2_params(fa2_params, k)?;

    let res = forceatlas2(
        data,
        pre_computed_knn,
        &fa2_params,
        FA2_APPROX,
        seed,
        verbose,
    )
    .to_extendr()?;

    Ok(Mat::from_fn(res[0].len(), 2, |i, j| res[j][i]))
}

/// Wrapper function into ForceAtlas2 on a caller-supplied graph
///
/// The edge list comes from an undirected igraph object, which stores every
/// edge once. The crate needs both directions, so each edge is mirrored here.
/// Self-loops are kept once; the crate drops them.
///
/// ### Params
///
/// * `from` - 1-based source vertex of each edge.
/// * `to` - 1-based target vertex of each edge.
/// * `weight` - Weight of each edge.
/// * `n` - Number of vertices.
/// * `init` - Optional initial layout as the column-major data of an `n x 2`
///   R matrix. `None` gives a random layout.
/// * `fa2_params` - Named R list with the ForceAtlas2 parameters.
/// * `seed` - For reproducibility.
/// * `verbose` - If `0` -> silent or `1` for normal verbosity, `2` for detailed
///   verbosity.
///
/// ### Returns
///
/// The ForceAtlas2 embedding as samples x 2 matrix.
///
/// ### References
///
/// Jacomy et al., PLoS ONE, 2014
#[allow(clippy::too_many_arguments)]
pub fn fa2_graph_manifold<T>(
    from: &[i32],
    to: &[i32],
    weight: &[f64],
    n: usize,
    init: Option<&[f64]>,
    fa2_params: List,
    seed: usize,
    verbose: usize,
) -> Result<Mat<T>, extendr_api::Error>
where
    T: ManifoldsFloat,
{
    let optim_params = get_params_fa2_optim(fa2_params)?;

    let n_entries = 2 * from.len();
    let mut row_indices = Vec::with_capacity(n_entries);
    let mut col_indices = Vec::with_capacity(n_entries);
    let mut values = Vec::with_capacity(n_entries);

    for ((&a, &b), &w) in from.iter().zip(to).zip(weight) {
        let (a, b) = ((a - 1) as usize, (b - 1) as usize);
        let w = T::from_f64(w).unwrap();
        row_indices.push(a);
        col_indices.push(b);
        values.push(w);
        if a != b {
            row_indices.push(b);
            col_indices.push(a);
            values.push(w);
        }
    }

    let graph = CoordinateList {
        row_indices,
        col_indices,
        values,
        n_samples: n,
    };

    let init = init.map(|flat| {
        flat.chunks(n)
            .map(|col| col.iter().map(|&v| T::from_f64(v).unwrap()).collect())
            .collect()
    });

    let res = forceatlas2_from_graph(&graph, init, &optim_params, FA2_APPROX, seed, verbose)
        .to_extendr()?;

    Ok(Mat::from_fn(n, 2, |i, j| res[j][i]))
}
